src/js/common/utils/accessible-menus.js

```diff
@@ -0,0 +1,349 @@
+  //  Up, down, left, right, enter (if necessary), space (if necessary)
+  menuElement.addEventListener('keydown', (event) => {
+    const targetLi = event.target.parentElement;
+    // Target's grandparent because the parent is a <li>
+    const inMenubar = targetLi.parentElement.getAttribute('role').toLowerCase() === 'menubar';
+    const inMenu = targetLi.parentElement.getAttribute('role').toLowerCase() === 'menu';
+
+    switch (event.keyCode) {
+      case LEFT_ARROW: {
+        if (inMenubar) {
+          event.preventDefault();
+          closeMenu(targetLi);
+          moveFocus(targetLi, 'previous');
+        } else if (inMenu) {
+          event.preventDefault();
+
+          // Move focus to the opening menu button
+          // This wins no prizes for efficiency, but until we have efficiency problems...
+          // Find the nearest ancestor who has a menu structure
+          const parentMenuLi = findNearestAncestor(targetLi, getMenuStructure);
+          firstVisibleChild(parentMenuLi).focus();
+        }
+        break;
+      }
+
+      case RIGHT_ARROW: {
+        if (inMenubar) {
+          event.preventDefault();
+          closeMenu(targetLi);
+          moveFocus(targetLi, 'next');
+        } else if (isMenuButton(event.target)) {
+          event.preventDefault();
+          // Open the menu, focus on the first item
+          openMenu(targetLi);
+        }
+        break;
+      }
+
+      case UP_ARROW: {
+        const isMB = isMenuButton(event.target);
+        if (inMenubar && isMB) {
+          // Open the menu, focus on the last item
+          event.preventDefault();
+
+          // Only open the first menu; we'll manually open the next level
+          openMenu(targetLi);
+
+          if (isWideScreen()) {
+            // `[role="menu"] > li:last-child` was acting more like `[role="menu"] > li :last-child`, so we're
+            //  doing it the hard way
+            const menuItems = targetLi.querySelector('[role="menu"]').children;
+            const lastSubmenu = menuItems[menuItems.length - 1];
+            openMenu(lastSubmenu, false, false);
+            lastSubmenu.firstElementChild.focus();
+          }
+        } else if (inMenu) {
+          event.preventDefault();
+          // If we've reached the top of a list, figure out if we need to go up a level or wrap around
+          if (targetLi.previousElementSibling === null) {
+            const firstLi = findNearestLi(targetLi);
+
+            // If the first <li> ancestor is in a [role="menubar"], we're at the first submenu and should
+            //  go up to the top-level menubar item
+            if (firstLi && (firstLi.parentElement.getAttribute('role') || '') === 'menubar') {
+              firstLi.firstElementChild.focus();
+            } else {
+              moveFocus(targetLi, 'previous');
+            }
+          } else {
+            moveFocus(targetLi, 'previous');
+          }
+        }
+        break;
+      }
+
+      case DOWN_ARROW: {
```

Comment:
I am not familiar with the expected behavior of mega menus, if this opens a menu, should space and enter do the same ([example 1](http://adobe-accessibility.github.io/Accessible-Mega-Menu/), [example 2](https://dequeuniversity.com/library/aria/menus/megamenu#))?
It also looks like [some examples](https://staff.washington.edu/tft/tests/menus/simplyaccessible/index.html)  open as soon as a menu option receives focus, so it's not clear that there is a well-established standard (but you probably found more helpful resources), though the first two examples seem pretty solid.
It is interesting to note the use of anchor tags in the Deque and W3 examples, as this sort of duplicate event handling can be simplified by using click (as according to the W3 example: "Despite the name, click events are activated regardless of the input method as soon as the link gets activated.")